kroxylicious-filters/kroxylicious-record-validation/src/test/java/io/kroxylicious/proxy/filter/schema/validation/topic/PerRecordTopicValidatorTest.java

```diff
@@ -0,0 +1,149 @@
+                .isEmpty();
+
+        verify(this.validator, times(1)).validate(any(Record.class));
+    }
+
+    @Test
+    void singleInvalidRecord() {
+        var tpd = createTopicProduceDataWithOneRecord(RecordTestUtils.record("bad"));
+        var topicValidator = new PerRecordTopicValidator(validator);
+
+        when(validator.validate(any(Record.class))).thenReturn(CompletableFuture.completedStage(new Result(false, "my bad record")));
+        var result = topicValidator.validateTopicData(tpd);
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .returns(true, TopicValidationResult::isAllPartitionsInvalid)
+                .returns(true, TopicValidationResult::isAnyPartitionInvalid);
+
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .extracting(TopicValidationResult::invalidPartitions, stream(PartitionValidationResult.class))
+                .singleElement()
+                .extracting(PartitionValidationResult::recordValidationFailures, list(RecordValidationFailure.class))
+                .singleElement()
+                .extracting(RecordValidationFailure::errorMessage, STRING)
+                .contains("my bad record");
+
+        verify(this.validator, times(1)).validate(any(Record.class));
+    }
+
+    @Test
+    void mixedResultWithinSinglePartition() {
+        var good = RecordTestUtils.record(0, null, "good");
+        var bad = RecordTestUtils.record(1, null, "bad");
+        var tpd = createTopicProduceDataWithOneRecord(good, bad);
+        var topicValidator = new PerRecordTopicValidator(validator);
+
+        when(validator.validate(good)).thenReturn(Result.VALID);
+        when(validator.validate(bad)).thenReturn(CompletableFuture.completedStage(new Result(false, "my bad record")));
+        var result = topicValidator.validateTopicData(tpd);
+
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .extracting(TopicValidationResult::invalidPartitions, stream(PartitionValidationResult.class))
+                .singleElement()
+                .extracting(PartitionValidationResult::recordValidationFailures, list(RecordValidationFailure.class))
+                .singleElement()
+                .returns("my bad record", RecordValidationFailure::errorMessage)
+                .returns(1, RecordValidationFailure::invalidIndex);
+
+        verify(this.validator, times(2)).validate(any(Record.class));
+    }
+
+    @Test
+    void mixedResultAcrossTwoPartitions() {
+        var good = RecordTestUtils.record("good");
+        var bad = RecordTestUtils.record(0, null, "bad");
+        var ugly = RecordTestUtils.record(1, null, "ugly");
+
+        var tpd = new TopicProduceData();
+        tpd.partitionData().add(makePartitionProduceData(0, good));
+        tpd.partitionData().add(makePartitionProduceData(1, bad, ugly));
```

Comment:
is it worth adding a test for 
```suggestion
        var tpd = new TopicProduceData();
        tpd.partitionData().add(makePartitionProduceData(0, good, bad));
        tpd.partitionData().add(makePartitionProduceData(1, ugly));
```

So that all partitions have some bad data?

Reply:
Done